src/main/java/appeng/fluids/storage/FluidCellInventory.java

```diff
@@ -0,0 +1,285 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.fluids.storage;
+
+
+import java.util.HashSet;
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.fluids.FluidStack;
+
+import appeng.api.AEApi;
+import appeng.api.config.Actionable;
+import appeng.api.exceptions.AppEngException;
+import appeng.api.implementations.items.IStorageCell;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.storage.IMEInventoryHandler;
+import appeng.api.storage.ISaveProvider;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IFluidStorageChannel;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.core.AEConfig;
+import appeng.core.AELog;
+import appeng.me.storage.CellInventoryBase;
+import appeng.me.storage.CellInventoryHandler;
+import appeng.util.item.AEFluidStack;
+
+
+/**
+ * @author DrummerMC
+ * @version rv6 - 2018-01-16
+ * @since rv6 2018-01-16
+ */
+public class FluidCellInventory extends CellInventoryBase<IAEFluidStack>
+{
+	private static final HashSet<String> BLACK_LIST = new HashSet<>();
+
+	protected FluidCellInventory( final NBTTagCompound data, final ISaveProvider container )
+	{
+		super( data, container, 8000 );
+	}
+
+	private FluidCellInventory( final ItemStack o, final ISaveProvider container ) throws AppEngException
+	{
+		super( o, container, 8000 );
+	}
+
+	public static IMEInventoryHandler getCell( final ItemStack o, final ISaveProvider container2 )
+	{
+		try
+		{
+			return new CellInventoryHandler( new FluidCellInventory( o, container2 ), AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) );
+		}
+		catch( final AppEngException e )
+		{
+			return null;
+		}
+	}
+
+	public static boolean isCell( final ItemStack i )
+	{
+		if( i == null )
+		{
+			return false;
+		}
+
+		final Item type = i.getItem();
+		if( type instanceof IStorageCell )
+		{
+			if( ( (IStorageCell) type ).getChannel() == AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) )
+			{
+				return ( (IStorageCell) type ).isStorageCell( i );
+			}
+		}
+
+		return false;
+	}
+
+	public static void addBasicBlackList( final String fluidName )
+	{
+		BLACK_LIST.add( fluidName );
+	}
+
+	private static boolean isBlackListed( final IAEFluidStack input )
+	{
+		return BLACK_LIST.contains( input.getFluid().getName() );
+	}
+
+	@Override
+	public IAEFluidStack injectItems( final IAEFluidStack input, final Actionable mode, final IActionSource src )
+	{
+		if( input == null )
+		{
+			return null;
+		}
+		if( input.getStackSize() == 0 )
+		{
+			return null;
+		}
+
+		if( isBlackListed( input ) || this.cellType.isBlackListed( this.i, input ) )
+		{
+			return input;
+		}
+
+		final FluidStack sharedFluidStack = input.getFluidStack();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Move the declaration of "sharedFluidStack" closer to the code that uses it. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1941)